File: py_test/integration/test_wasm_middleware.py
"""Integration tests for WASM OnRequest middleware (real Router + mock worker)."""

from __future__ import annotations

import subprocess
from pathlib import Path

import pytest
import requests

REPO_ROOT = Path(__file__).resolve().parents[2]
EXAMPLE_DIR = REPO_ROOT / "examples" / "wasm_middleware"
WASM_COMPONENT = EXAMPLE_DIR / "wasm_middleware_example.component.wasm"


def _router_bin() -> Path:
    for candidate in (
        REPO_ROOT / "target" / "release" / "vllm-router",
        REPO_ROOT / "target" / "debug" / "vllm-router",
    ):
        if candidate.is_file():
            return candidate
    pytest.skip(
        "vllm-router binary not found; build with "
        "`cargo build --release --bin vllm-router` first"
    )


def _ensure_wasm_component() -> Path:
    if WASM_COMPONENT.is_file():
        return WASM_COMPONENT
    build_sh = EXAMPLE_DIR / "build.sh"
    if not build_sh.is_file():
        pytest.skip(f"missing example plugin build script at {build_sh}")
    subprocess.check_call(["bash", str(build_sh)], cwd=EXAMPLE_DIR)
    if not WASM_COMPONENT.is_file():
        pytest.skip(f"failed to build WASM component at {WASM_COMPONENT}")
    return WASM_COMPONENT


def _header_map(payload: dict) -> dict:
    headers = payload.get("request_headers") or {}
    return {str(k).lower(): v for k, v in headers.items()}


@pytest.mark.integration
def test_wasm_middleware_modify_reject_and_path_isolation(router_manager, mock_workers):
    """Example plugin: set header on chat, reject marker body, skip other routes."""
    wasm_path = _ensure_wasm_component()
    router_bin = _router_bin()
    _, urls, _ = mock_workers(n=1)

    rh = router_manager.start_router(
        worker_urls=urls,
        policy="round_robin",
        router_bin=str(router_bin),
        extra={
            "worker_startup_timeout_secs": 30,
            "wasm_middleware": str(wasm_path),
        },
    )

    # Modify: chat should forward with the example header.
    chat = requests.post(
        f"{rh.url}/v1/chat/completions",
        json={
            "model": "mock",
            "messages": [{"role": "user", "content": "hi"}],
        },
        timeout=30,
    )
    assert chat.status_code == 200, chat.text
    chat_headers = _header_map(chat.json())
    assert chat_headers.get("x-wasm-middleware") == "example", chat_headers

    # Reject: fail closed at the Router before the worker.
    rejected = requests.post(
        f"{rh.url}/v1/chat/completions",
        json={
            "model": "mock",
            "messages": [{"role": "user", "content": "nope"}],
            "note": "__wasm_reject__",
        },
        timeout=30,
    )
    assert rejected.status_code == 400, rejected.text

    # Path isolation: completions is not attached by default.
    completions = requests.post(
        f"{rh.url}/v1/completions",
        json={"model": "mock", "prompt": "hi"},
        timeout=30,
    )
    assert completions.status_code == 200, completions.text
    completion_headers = _header_map(completions.json())
    assert "x-wasm-middleware" not in completion_headers, completion_headers

File: tests/test_wasm_middleware.rs
//! Integration tests for WASM OnRequest middleware with mock HTTP workers.
//!
//! Requires the example component artifact:
//! `bash examples/wasm_middleware/build.sh`

mod common;

use axum::{
    body::Body,
    extract::Request,
    http::{header::CONTENT_TYPE, StatusCode},
};
use common::mock_worker::{
    self, HealthStatus, MockWorker, MockWorkerConfig, WorkerType,
};
use common::test_app::create_test_app_with_wasm;
use http_body_util::BodyExt;
use reqwest::Client;
use serde_json::json;
use std::path::PathBuf;
use std::sync::Arc;
use tower::ServiceExt;
use vllm_router_rs::config::{RouterConfig, RoutingMode};
use vllm_router_rs::routers::RouterFactory;
use vllm_router_rs::wasm_middleware::{
    example_component_artifact_path, WasmMiddlewareConfig, WasmMiddlewareRuntime,
};

fn require_component_path() -> PathBuf {
    example_component_artifact_path().unwrap_or_else(|| {
        panic!(
            "WASM component artifact missing; run \
             `examples/wasm_middleware/build.sh` before \
             `cargo test --test test_wasm_middleware`"
        )
    })
}

fn header_values<'a>(
    captured: &'a mock_worker::CapturedRequest,
    name: &str,
) -> Option<&'a Vec<String>> {
    captured
        .headers
        .iter()
        .find(|(k, _)| k.eq_ignore_ascii_case(name))
        .map(|(_, v)| v)
}

#[tokio::test]
async fn wasm_middleware_modify_reject_and_path_isolation() {
    let component = require_component_path();
    let wasm_runtime = Arc::new(
        WasmMiddlewareRuntime::load(WasmMiddlewareConfig::from_path(component))
            .expect("load wasm runtime"),
    );

    let mut worker = MockWorker::new(MockWorkerConfig {
        port: 0,
        worker_type: WorkerType::Regular,
        health_status: HealthStatus::Healthy,
        response_delay_ms: 0,
        fail_rate: 0.0,
    });
    let worker_url = worker.start().await.expect("start mock worker");
    let worker_port: u16 = worker_url
        .rsplit(':')
        .next()
        .unwrap()
        .parse()
        .expect("parse worker port");
    mock_worker::clear_captured_requests(worker_port);

    let config = RouterConfig {
        mode: RoutingMode::Regular {
            worker_urls: vec![worker_url],
        },
        worker_startup_timeout_secs: 1,
        worker_startup_check_interval_secs: 1,
        ..Default::default()
    };
    let app_context = common::create_test_context(config.clone());
    let router = RouterFactory::create_router(&app_context)
        .await
        .expect("create router");
    tokio::time::sleep(tokio::time::Duration::from_millis(300)).await;

    let app = create_test_app_with_wasm(
        Arc::from(router),
        Client::new(),
        &config,
        false,
        Some(wasm_runtime),
    );

    // Modify: chat should forward with the example header.
    let chat = app
        .clone()
        .oneshot(
            Request::builder()
                .method("POST")
                .uri("/v1/chat/completions")
                .header(CONTENT_TYPE, "application/json")
                .body(Body::from(
                    json!({
                        "model": "mock-model",
                        "messages": [{"role": "user", "content": "hi"}]
                    })
                    .to_string(),
                ))
                .unwrap(),
        )
        .await
        .unwrap();
    assert_eq!(chat.status(), StatusCode::OK);
    let _ = chat.into_body().collect().await.unwrap().to_bytes();

    let captured = mock_worker::get_captured_requests(worker_port);
    let chat_req = captured
        .iter()
        .find(|r| r.path == "/v1/chat/completions")
        .expect("mock worker should receive chat request");
    let marker = header_values(chat_req, "x-wasm-middleware")
        .and_then(|v| v.first())
        .map(String::as_str);
    assert_eq!(marker, Some("example"));

    // Reject: fail closed at the Router before the worker.
    mock_worker::clear_captured_requests(worker_port);
    let rejected = app
        .clone()
        .oneshot(
            Request::builder()
                .method("POST")
                .uri("/v1/chat/completions")
                .header(CONTENT_TYPE, "application/json")
                .body(Body::from(
                    json!({
                        "model": "mock-model",
                        "messages": [{"role": "user", "content": "nope"}],
                        "note": "__wasm_reject__"
                    })
                    .to_string(),
                ))
                .unwrap(),
        )
        .await
        .unwrap();
    assert_eq!(rejected.status(), StatusCode::BAD_REQUEST);
    assert!(
        mock_worker::get_captured_requests(worker_port).is_empty(),
        "rejected requests must not reach the worker"
    );

    // Path isolation: completions is not attached by default.
    mock_worker::clear_captured_requests(worker_port);
    let completions = app
        .oneshot(
            Request::builder()
                .method("POST")
                .uri("/v1/completions")
                .header(CONTENT_TYPE, "application/json")
                .body(Body::from(
                    json!({
                        "model": "mock-model",
                        "prompt": "hi"
                    })
                    .to_string(),
                ))
                .unwrap(),
        )
        .await
        .unwrap();
    assert_eq!(completions.status(), StatusCode::OK);
    let _ = completions.into_body().collect().await.unwrap().to_bytes();

    let captured = mock_worker::get_captured_requests(worker_port);
    let completion_req = captured
        .iter()
        .find(|r| r.path == "/v1/completions")
        .expect("mock worker should receive completions request");
    assert!(
        header_values(completion_req, "x-wasm-middleware").is_none(),
        "default WASM attach point must not modify /v1/completions"
    );

    worker.stop().await;
}
